src/db.rs

```diff
@@ -0,0 +1,590 @@
+//! A sqlite based block store for content-addressed data that tries to do as much as possible
+//! in the database.
+//!
+//! This module is for all interactions with the database, so all SQL statements go in here.
+//!
+//! Tables:
+//! cids: mapping from cid (blob < 64 bytes) to id (u64)
+//! refs: m:n mapping from block ids to their children
+//! blocks: the actual data for blocks, keyed by block id
+//!    cids can exist in the system without having data associated with them!
+//! alias: table that contains named pins for roots of graphs that should not be deleted by gc
+//!    you can alias incomplete or in fact non-existing data. It is not necessary for a pinned dag
+//!    to be complete.
+use libipld::{Cid, DefaultParams};
+use rusqlite::{
+    config::DbConfig, params, types::FromSql, Connection, OptionalExtension, ToSql, Transaction,
+    NO_PARAMS,
+};
+use std::{
+    collections::BTreeSet,
+    convert::TryFrom,
+    sync::atomic::{AtomicI64, Ordering},
+    time::Duration,
+    time::Instant,
+};
+use tracing::*;
+
+use crate::{cache::CacheTracker, SizeTargets, StoreStats};
+
+const PRAGMAS: &str = r#"
+PRAGMA foreign_keys = ON;
+PRAGMA journal_mode = WAL;
+-- PRAGMA synchronous = NORMAL;
+PRAGMA synchronous = FULL;
+PRAGMA page_size = 4096;
+-- PRAGMA locking_mode = EXCLUSIVE;
```

Comment:
Why would we ever want `EXCLUSIVE`? Perhaps remove this comment?

Reply:
It was a test to see if it makes things go faster. It does not, at least I was unable to measure a difference.